lib/js/src/manager/screen/BaseVoiceCommandManager.js

```diff
@@ -0,0 +1,219 @@
+import { AddCommand } from '../../rpc/messages/AddCommand.js';
+
+class BaseVoiceCommandManager extends BaseSubManager {
+    /**
+     * @param {LifecycleManager} lifecycleManager
+    */
+    constructor (lifecycleManager) {
+        super(lifecycleManager);
+        this._currentHmiLevel = HMILevel.HMI_NONE;
+        this._taskQueue = [];
+        this._voiceCommands = [];
+        this._voiceCommandIdMin = 1900000000;
+        this._lastVoiceCommandId = this._voiceCommandIdMin;
+        this._isReady = false; // whether the HMI level is appropriate for sending requests
+        this._hmiListener = null;
+        this._commandListener = null;
+        this._isUpdateRunning = false; // prevent _update from being invoked more than once at a time
+        this._addListeners();
+    }
+
+    /**
+     * @return {Promise}
+    */
+    async start () {
+        this._transitionToState(BaseSubManager.READY);
+        super.start();
+    }
+
+    dispose () {
+        this._currentHmiLevel = HMILevel.HMI_NONE;
+        this._taskQueue = [];
+        this._voiceCommands = [];
+        this._lastVoiceCommandId = this._voiceCommandIdMin;
+        this._isReady = false;
+        // remove listeners
+        this._lifecycleManager.removeRpcListener(FunctionID.ON_HMI_STATUS, this._hmiListener);
+        this._lifecycleManager.removeRpcListener(FunctionID.ON_COMMAND, this._commandListener);
+        this._hmiListener = null;
+        this._commandListener = null;
+        this._isUpdateRunning = false;
+
+        super.dispose();
+    }
+
+    /**
+     * @param {VoiceCommand[]} voiceCommands
+     * @return {Promise} - returns after old commands are deleted and new ones are added
+    */
+    async setVoiceCommands (voiceCommands) {
+        // we actually need voice commands to set.
+        if (!Array.isArray(voiceCommands)) {
+            return;
+        }
+        // add the commands to a queue to be processed later
+        this._taskQueue.push(voiceCommands);
+
+        if (!this._isUpdateRunning) {
+            this._isUpdateRunning = true;
+            await this._update();
+            this._isUpdateRunning = false;
+        }
+    }
+
+    /**
+     * Looks through the task queue to process incoming voice commands. Recursive
+     * @return {Promise} - returns after old commands are deleted and new ones are added
+    */
+    async _update () {
+        // we actually need voice commands to set.
+        if (this._taskQueue.length === 0) {
+            return;
+        }
+
+        if (!this._isReady) { // not in the right HMI level yet
+            return;
+        }
+
+        // remove all but the last element in the task queue for processing
+        // there is only going to be one voice command set in the end
+        // any other than the last element will end up being deleted anyway
+        let voiceCommands = [];
+        while (this._taskQueue.length > 0) {
+            voiceCommands = this._taskQueue.shift();
+        }
```

Comment:
Would `.splice(...)` perform better than a while loop with `.shift()`?

Reply:
Will update to use splice